pkg/controllers/cronfederatedhpa/cronfederatedhpa_controller.go

```diff
@@ -0,0 +1,200 @@
+		if apierrors.IsNotFound(err) {
+			klog.V(4).Infof("Begin to cleanup the cron jobs for CronFederatedHPA:%s", req.NamespacedName)
+			c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+			return controllerruntime.Result{}, nil
+		}
+
+		klog.Error("Failed to query CronFederatedHPA(%s):", req.NamespacedName.String(), err)
+		return controllerruntime.Result{Requeue: true}, err
+	}
+
+	var err error
+	defer func() {
+		if err != nil {
+			c.EventRecorder.Event(cronFHPA, corev1.EventTypeWarning, "ProcessCronFederatedHPAFailed", err.Error())
+		}
+	}()
+
+	if !cronFHPA.DeletionTimestamp.IsZero() {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+		return controllerruntime.Result{}, nil
+	}
+
+	origRuleMap := make(map[string]struct{})
+	for _, history := range cronFHPA.Status.ExecutionHistories {
+		origRuleMap[history.RuleName] = struct{}{}
+	}
+
+	// If scale target is updated, stop all the rule executors
+	if c.CronHandler.CronFHPAScaleTargetRefUpdates(req.NamespacedName.String(), cronFHPA.Spec.ScaleTargetRef) {
+		c.CronHandler.StopCronFHPAExecutor(req.NamespacedName.String())
+	}
+
+	c.CronHandler.AddCronExecutorIfNotExist(req.NamespacedName.String())
+
+	newRuleMap := make(map[string]struct{})
+	for _, rule := range cronFHPA.Spec.Rules {
+		if err = c.processCronRule(cronFHPA, req.NamespacedName, rule); err != nil {
+			return controllerruntime.Result{}, err
+		}
+		newRuleMap[rule.Name] = struct{}{}
+	}
+
+	// If rule is deleted, remove the rule executor from the handler
+	for name := range origRuleMap {
+		if _, ok := newRuleMap[name]; !ok {
+			c.CronHandler.StopRuleExecutor(req.NamespacedName.String(), name)
+			if err = c.removeCronFHPAHistory(cronFHPA, name); err != nil {
+				return controllerruntime.Result{}, err
+			}
+		}
+	}
+
+	return controllerruntime.Result{}, nil
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CronFHPAController) SetupWithManager(mgr controllerruntime.Manager) error {
+	c.CronHandler = NewCronHandler(mgr.GetClient(), mgr.GetEventRecorderFor(ControllerName))
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&autoscalingv1alpha1.CronFederatedHPA{}).
+		WithOptions(controller.Options{RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RateLimiterOptions)}).
+		Complete(c)
+}
+
+// processCronRule processes the cron rule
+func (c *CronFHPAController) processCronRule(cronFHPA *autoscalingv1alpha1.CronFederatedHPA,
+	namespaceName types.NamespacedName, rule autoscalingv1alpha1.CronFederatedHPARule) error {
+	ruleOld, exists := c.CronHandler.RuleCronExecutorExists(namespaceName.String(), rule.Name)
+	if exists {
+		if !equality.Semantic.DeepEqual(ruleOld, rule) {
+			c.CronHandler.StopRuleExecutor(namespaceName.String(), rule.Name)
+		} else {
+			// rule is not changed, ignore it
+			return nil
+		}
+	}
+
+	if helper.IsCronFederatedHPARuleSuspend(rule) {
```

Comment:
"Suspend" not only has this meaning, but it also means to suspend an existing cron job.